bluesky/bluesky
# Propagate exception through failed status

It was noticed that ophyd v2 AsyncStatus objects do not carry their exceptions through to the run engine. This relates to the following issue: https://github.com/bluesky/ophyd/issues/1098

## Description
There are several small changes involved in this PR:
1. Status protocol now has an 'exception' property, which can return exceptions.
2. When a FailedStatus is created (in `RunEngine::_status_object_completed`) this method now checks if the status has exception as a method (this is for backwards compatibility with ophyd v1 devices). It retrieves the exception and raises a FailedStatus from it.

These have been tested against a PR in the ophyd library, found here: https://github.com/bluesky/ophyd/pull/1103

## Motivation and Context
It is annoying for users to have bluesky plans fail with ophyd v2 devices, and get no more information than a simple FailedStatus. In addition, the FailedStatus is often one of the most recent calls in traceback, so it appears first visually, and seeing something like:

```
ERROR FailedStatus: None
```

can be a bit disheartening.

The aim is, that this PR will help to propagate errors through the FailedStatus in ophyd v2 devices.
## How Has This Been Tested?
I have added a couple of tests, one of which will fail until the ophyd v2 pull request has been merged. These just check how ophyd devices propagate their errors - for the ophyd v2 case, one can easily test the implications of the changes in these PRs:

1. Clone this repository
2. install requirements, i.e. `pip install -e . / pip install -r requirements-dev.txt`
3. install ophyd, i.e. `pip install ophyd`
4. `pytest bluesky/tests/test_run_engine.py -k test_failures_propagated_through_FailedStatus_ophyd_v2`
5. Observe the output; the test will fail, and in the failure the FailedStatus will contain no more information.
6. uninstall ophyd and reinstall from the PR mentioned above by typing, `pip install git+https://github.com/bluesky/ophyd.git@propagate_exception_through_FailedStatus`
7. repeat steps 5 and 6, noticing the tests pass as the FailedStatus now recognizes the correct exception.
